app/reporter/reporter.go

```diff
@@ -76,16 +101,41 @@ func (l Reporter) activate() {
 	for {
 		select {
 		case entry := <-l.messages:
-			buffer = append(buffer, entry)
+			// don't save right away, wait for antispam checks
+			time.Sleep(l.saveDelay)
+
+			if !l.messageExists(entry.MessageID) {
+				log.Printf("[DEBUG] message %d has been deleted, skipping", entry.MessageID)
+				continue
+			}
+
+			buffer = append(buffer, entry.Data)
 			if len(buffer) >= 100 { // forced flush every 100 records
 				if err := writeBuff(); err != nil {
 					log.Printf("[WARN] failed to write reporter buffer, %v", err)
 				}
 			}
-		case <-time.After(time.Second * 5): // flush on 5 seconds inactivity
+		case <-ticker.C: // flush on 5 seconds inactivity
 			if err := writeBuff(); err != nil {
 				log.Printf("[WARN] failed to write reporter buffer, %v", err)
 			}
 		}
 	}
 }
+
+//go:generate moq -out mock_reporter.go . httpClient
+
+type httpClient interface {
```

Comment:
nit: I usually expect interfaces to be defined in the top section, right after the primary struct definitions.